File: Data Structures & Algorithms/trapping-rain-water/submission-4.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if (height.empty()) return 0;
        int left = 0, right = height.size() - 1;
        int leftMax = height[left], rightMax = height[right];
        int trap_area = 0;
        while (left < right) {
            if (height[left] < height[right]) {
                ++left;
                leftMax = max(leftMax, height[left]);
                trap_area += leftMax - height[left];
            } else {
                --right;
                rightMax = max(rightMax, height[right]);
                trap_area += rightMax - height[right];
            }
        }
        return trap_area;
    }
};

File: Data Structures & Algorithms/valid-parenthesis-string/submission-4.go
func checkValidString(s string) bool {
	// the min / max count of possible unmatched '('
	leftMin, leftMax := 0, 0
	for i := range s {
		if s[i] == '(' {
			leftMin++
			leftMax++
		} else if s[i] == ')' {
			leftMin--
			leftMax--
		} else {
			leftMin--
			leftMax++
		}
		if leftMax < 0 {
			// too more right parenthesis
			return false
		}
		if leftMin < 0 {
			// we seem too more stars as the right parenthesis,
			// so we reset star to an empty string.
			leftMin = 0
		}
	}
	return leftMin == 0
}